### PR TITLE
Fix grid missing border patches

### LINKED ISSUE #5: Missing patches when zarr-chunks is not "full"
Related to #3 (Inference Sampler)

The grid created by [`zds.PatchSampler`](https://github.com/TheJacksonLaboratory/zarrdataset/blob/fb57026a1ed7436603d66b440ab4f9ebfee377f9/zarrdataset/_samplers.py#L13) doesn't take into account border of zarr if the zarr.shape is not a multiple of chunk.shape

Small example:
```python
%load_ext autoreload
%autoreload 2

import zarr
import zarrdataset as zds
from torch.utils.data import DataLoader

filename = r"data.zarr"

# create empty zarr dataset
z = zarr.zeros((1, 1, 6), chunks=(1, 1, 4), dtype='uint8')
zarr.save(filename, z)


patch_size = dict(Z=1, Y=1, X=2)
patch_sampler = zds.PatchSampler(patch_size=patch_size)

my_datasets = zds.ZarrDataset(
    [
    zds.ImagesDatasetSpecs(
        filenames=filename,
        source_axes="ZYX",
        axes="ZYX",
    )
    ],
    patch_sampler=patch_sampler,
    return_positions=True,
    return_worker_id=True
)

my_dataloader = DataLoader(my_datasets,
                    num_workers=0,
                        worker_init_fn=zds.zarrdataset_worker_init_fn,
                    batch_size=1
                    )

for i, (wid, pos, sample) in enumerate(my_dataloader):
    print(pos)

```

result:

```python
tensor([[[0, 1],
         [0, 1],
         [0, 2]]])
tensor([[[0, 1],
         [0, 1],
         [2, 4]]])
```

Is there a reason it doesn't return (or is it a bug?)

```python
tensor([[[0, 1],
         [0, 1],
         [0, 2]]])
tensor([[[0, 1],
         [0, 1],
         [2, 4]]])
tensor([[[0, 1],
         [0, 1],
         [4, 6]]])
```

### PR DESCRIPTION
**Pull request by ClementCaporal**
Attempt to close #5 

This change allows `zarrdataset` to extract all samples even at the border of the image when chunk-zarr is not fully initialized..
This is helpful when using zarrdataset for inference on larger-than-memory inputs.

Potential problems:
- I am not sure to understand why https://github.com/TheJacksonLaboratory/zarrdataset/blob/fb57026a1ed7436603d66b440ab4f9ebfee377f9/zarrdataset/_samplers.py#L74-L91 has to compute several factors. Thus I selected only the first factor when using `corners_idx` here https://github.com/LaboratoryOpticsBiosciences/zarrdataset/blob/0fe3403dc29d99770cea59438e6b05d75438175c/zarrdataset/_samplers.py#L182 and https://github.com/LaboratoryOpticsBiosciences/zarrdataset/blob/0fe3403dc29d99770cea59438e6b05d75438175c/zarrdataset/_samplers.py#L189 (Would it cause problems?)
- Some part of the image might be loaded twice. As work is distrbuted to several workers over chunks, some sample can be present in several chunks.

### CODE REVIEW

**Comment by fercer** [comment]
Thanks @ClementCaporal, I can briefly explain about the factors there (and probably put this in the documentation too). 

The main idea is that ZarrDataset can extract patches from masked locations. And that mask can be of different scale than the input image, e.g. one generated from a down sampled version of the input image.
So, the factors there are to check what pixels/voxels are inside masked locations. Because Zarr arrays are n-dimensional the check process takes into account all vertices in a n-dimensional box.

The issue was indeed originated in the sections of the code that you modified.
I integrated these fixes to this problem in the last update of PR #4.

**Comment by ClementCaporal** [comment]
I propose to close this draft PR as it has been duplicated solved in #4 ?
